### DISCUSSION TITLE
How to replicate FreeText annots keeping text rotation

### DISCUSSION THREAD

**Discussion by izerui** [pymupdf/PyMuPDF#3338] [Q&A] [answered]
### Description of the bug

_annot = new_page.add_freetext_annot(rect=annot.rect,
                                                             text=annot.get_text(),
                                                             fontname=styles["font_name"],
                                                             fontsize=styles["font_size"],
                                                             text_color=styles["color"],
                                                             align=styles["text_align"],
                                                             rotate=rotation)

<img width="883" alt="企业微信截图_2db56ac6-9ae0-4aa5-ad35-ce28c8752557" src="https://github.com/pymupdf/PyMuPDF/assets/1773233/790cdd03-4482-426a-835f-43257cb42881">

### How to reproduce the bug
```
                if 'Rotation' in keys:
                    rotation = int(doc.xref_get_key(annot.xref, 'Rotation')[1])
                if rotation < 0:
                    rotation = rotation + 360
```

### PyMuPDF version

1.24.0

### Operating system

MacOS

### Python version

3.10

**Comment by JorjMcKie** [comment]
As always: We need an example file to reproduce - not just a picture!

**Comment by izerui** [comment]
https://tfile.yj2025.com/pdf-processor/source/2024-04-02/mt_04_24024_0_812-wps.pdf

**Comment by izerui** [comment]
test code like this:

```
    def test_bug_91(self):
        bytes = get_url_content_retry(
            'https://tfile.yj2025.com/pdf-processor/source/2024-04-02/mt_04_24024_0_812-wps.pdf')
        target_doc: Document = fitz.open()
        doc = fitz.open('pdf', bytes)
        for index, page in enumerate(doc):
            page.clean_contents()
            # print(doc.xref_object(page.xref))
            new_page = target_doc.new_page(width=page.rect.width, height=page.rect.height)
            new_page.set_rotation(page.rotation)
            # new_page.show_pdf_page(page.cropbox, doc, index, rotate=page.rotation, keep_proportion=True,
            #                        clip=page.cropbox)
            for annot_index, annot in enumerate(page.annots()):
                print('\n')
                print('Remote Control:', doc.xref_get_key(annot.xref, 'RC'))
                print('Default Style:', doc.xref_get_key(annot.xref, 'DS'))
                keys = doc.xref_get_keys(annot.xref)
                rotation = 0
                if 'Rotation' in keys:
                    rotation = int(doc.xref_get_key(annot.xref, 'Rotation')[1])
                if rotation < 0:
                    rotation = rotation + 360
                # 复制注释到目标文档中
                match annot.type[1]:
                    case 'FreeText':
                        _annot = new_page.add_freetext_annot(rect=annot.rect,
                                                             text=annot.get_text(),
                                                             rotate=rotation)
            pass
        target_doc.save('222.pdf', garbage=4, deflate=True)
```

**Comment by izerui** [comment]
get_url_content_retry  can be replace with:

```
httpx.get(url).content
```

**Comment by izerui** [comment]
This phenomenon also exists in version 1.24.1

**Comment by JorjMcKie** [comment]
If you follow the documentation, things work correctly, see here:
```python
doc=fitz.open("test.pdf")
page=doc[0]
page.add_freetext_annot((50,100,150,200),text,rotate=90,text_color=(1,0,0))
'FreeText' annotation on page 0 of test.pdf
doc.ez_save("x.pdf")
```
Delivers correctly rotated text including line breaks:
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/47d768aa-f752-4e7f-a11d-2588ff99a713)

**Comment by izerui** [comment]
Yes, if I add a annot at the agreed 90, the angle is correct, but the rotation of the original annot, obtained by xref, is -90. I would like to copy the annot as it is and use it in the new page, is there any way to correct the rotation obtained by xref?

**Comment by izerui** [comment]
> Yes, if I add a annot at the agreed 90, the angle is correct, but the rotation of the original annot, obtained by xref, is -90. I would like to copy the annot as it is and use it in the new page, is there any way to correct the rotation obtained by xref?

found through pymupdf added after the rotation of the annotations, and then get the xref keys is rotate, then why other tools to add the annotations through the xref to get the rotation is -90, and then annot.rotation directly is -1？


print(doc2.xref_get_keys(annot.xref))

way 1: page.add_freetext_annot((50,100,150,200),text,rotate=90,text_color=(1,0,0))
```
['Type', 'Subtype', 'Rect', 'BS', 'DA', 'P', 'F', 'Contents', 'Rotate', 'Q', 'AP', 'NM']
```

other tools way:
```
['Type', 'Subtype', 'AP', 'C', 'Contents', 'CreationDate', 'DA', 'DS', 'F', 'IT', 'M', 'NM', 'P', 'Rect', 'Rotation', 'Subj', 'T']
```

**Comment by izerui** [comment]
All of the text above is oriented at 90 degrees, but the results are not the same.

```
    def test_get_annot_rotation(self):
        bytes = httpx.get(
            'https://tfile.yj2025.com/pdf-processor/source/2024-04-03/x.pdf').content
        doc = fitz.open('pdf', bytes)
        page = doc[0]
        for annot in page.annots():
            print(annot.xref, annot.get_text(), doc.xref_get_keys(annot.xref))
            keys = doc.xref_get_keys(annot.xref)
            if 'Rotate' in keys:
                rotation = int(doc.xref_get_key(annot.xref, 'Rotate')[1])
                print('Rotate: ', rotation)
            if 'Rotation' in keys:
                rotation = int(doc.xref_get_key(annot.xref, 'Rotation')[1])
                print('Rotation: ', rotation)
```

output
```
 20 MT_04_24024_0_812
 ['Type', 'Subtype', 'AP', 'C', 'Contents', 'CreationDate', 'DA', 'DS', 'F', 'IT', 'M', 'NM', 'P', 'Rect', 'Subj', 'T']
21    MT_04_24024_0_812
 ['Type', 'Subtype', 'AP', 'C', 'Contents', 'CreationDate', 'DA', 'DS', 'F', 'IT', 'M', 'NM', 'P', 'Rect', 'Rotation', 'Subj', 'T']
Rotation:  -180
22 MS          CZ        CW
 ['Type', 'Subtype', 'AP', 'C', 'Contents', 'CreationDate', 'DA', 'DS', 'F', 'IT', 'M', 'NM', 'P', 'Rect', 'Subj', 'T']
23 2024-04-01
 ['Type', 'Subtype', 'AP', 'C', 'Contents', 'CreationDate', 'DA', 'DS', 'F', 'IT', 'M', 'NM', 'P', 'Rect', 'Subj', 'T']
24 数量：2（双台）
 ['Type', 'Subtype', 'AP', 'C', 'Contents', 'CreationDate', 'DA', 'DS', 'F', 'IT', 'M', 'NM', 'P', 'RC', 'Rect', 'Subj', 'T']
25 *41.455士0.1
 ['Type', 'Subtype', 'AP', 'C', 'Contents', 'CreationDate', 'DA', 'DS', 'F', 'IT', 'M', 'NM', 'P', 'RC', 'Rect', 'Rotation', 'Subj', 'T']
Rotation:  -90
26 20.728
 ['Type', 'Subtype', 'AP', 'C', 'Contents', 'CreationDate', 'DA', 'DS', 'F', 'IT', 'M', 'NM', 'P', 'Rect', 'Rotation', 'Subj', 'T']
Rotation:  -90
83 wps 注释
 ['Type', 'Subtype', 'P', 'Rect', 'F', 'Rotate', 'T', 'AP', 'NM', 'IT', 'DA', 'M', 'RC', 'DS', 'Contents']
Rotate:  0
92 some text with 
breaks
 ['Type', 'Subtype', 'Rect', 'BS', 'DA', 'P', 'F', 'Contents', 'Rotate', 'Q', 'AP', 'NM']
Rotate:  90
```

**Comment by JorjMcKie** [comment, marked as answer]
There are a number of alternatives how rotation of text in a FreeText annot can be achieved.
The only definitive way to find the actual text rotation is by using text extraction in "dict" format. The dict key "dir" in the line dictionary gives you the text orientation.
Use this (consult the [documentation](https://pymupdf.readthedocs.io/en/latest/textpage.html#line-dictionary)!) to compute the rotation angle to use when recreating the annot somewhere else.
In your case, you will find `"dir":(0,-1)` in the line dctionary.

**Comment by izerui** [comment]
Thanks, I'll look into line-dictionary.
